source/extensions/filters/common/attributes/attributes.h

```diff
@@ -0,0 +1,103 @@
+#pragma once
+
+#include <atomic>
+#include <chrono>
+#include <iterator>
+#include <map>
+#include <memory>
+#include <string>
+
+#include "absl/strings/str_format.h"
+#include "google/api/expr/v1alpha1/value.pb.h"
+
+#include "envoy/access_log/access_log.h"
+#include "envoy/buffer/buffer.h"
+#include "envoy/http/filter.h"
+#include "envoy/http/header_map.h"
+#include "envoy/stats/sink.h"
+#include "envoy/upstream/cluster_manager.h"
+#include "envoy/network/filter.h"
+
+#include "source/common/grpc/common.h"
+#include "source/common/http/headers.h"
+#include "source/common/protobuf/utility.h"
+#include "source/common/common/assert.h"
+#include "source/common/common/logger.h"
+#include "source/common/common/fmt.h"
+#include "source/common/common/lock_guard.h"
+#include "source/common/http/header_map_impl.h"
+#include "source/extensions/filters/common/attributes/id.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Filters {
+namespace Common {
+namespace Attributes {
+
+static const std::string HttpProtocolStrings[] = {"Http 1.0", "Http 1.1", "Http 2", "Http 3"};
+
+using HashPolicy = envoy::config::route::v3::RouteAction::HashPolicy;
+using MapValue = google::api::expr::v1alpha1::MapValue;
+using MapValue_Entry = google::api::expr::v1alpha1::MapValue_Entry;
+using Value = google::api::expr::v1alpha1::Value;
+using NullValue = google::protobuf::NullValue;
+using Any = google::protobuf::Any;
+
+class ValueUtil {
+public:
+  static Value* optionalStringValue(const absl::optional<std::string>& str);
+  static Value* stringValue(const std::string& str);
+  static Value* int64Value(int64_t n);
+  static Value* uint64Value(uint64_t n);
+  static Value* doubleValue(double n);
+  static Value* boolValue(bool b);
+  template <class T> static Value* objectValue(T val);
+  static Value* mapValue(MapValue* m);
+  static MapValue_Entry* getOrInsert(MapValue* m, absl::string_view key);
+  static Value* nullValue();
+};
+
+class Attributes : public Logger::Loggable<Logger::Id::filter> {
+public:
+  Attributes(StreamInfo::StreamInfo& stream_info) : stream_info_(stream_info){};
+
+  void setRequestHeaders(const Http::RequestHeaderMap* request_headers) {
```

Comment:
In general, what is the behavior of this library when the caller doesn't set all of these headers. For example, if the configuration requests a trailer attribute when the trailers aren't available, or requests a response header attribute on the request path, does the library just return no result, or does it crash? I think it's important that we don't crash in that situation and that we test for that case.

Reply:
Agreed it should not crash, my thinking was that those values should just be empty. I will write a test around this scenario specifically.